pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+
+	"github.com/grafana/mimir/pkg/storage/bucket"
+)
+
+// CreateBlockUpload requests the creation of a block upload session.
+func (c *MultitenantCompactor) CreateBlockUpload(ctx context.Context, tenantID, blockID string) error {
+	level.Info(c.logger).Log("msg", "creating block upload session", "user", tenantID, "block_id", blockID)
+	// TODO: Verify that block hasn't already been ingested
+	return nil
+}
+
+// UploadBlockFile uploads a block file.
+func (c *MultitenantCompactor) UploadBlockFile(ctx context.Context, tenantID, blockID, pth string, r *http.Request) error {
+	if path.Base(pth) == "meta.json" {
+		return fmt.Errorf("meta.json is not allowed")
+	}
+
+	level.Info(c.logger).Log("msg", "uploading block file", "user", tenantID,
+		"block_id", blockID, "path", pth, "size", r.ContentLength)
+
+	dst := path.Join(blockID, pth)
+	level.Info(c.logger).Log("msg", "uploading block file to bucket", "user", tenantID,
+		"destination", dst)
+	bkt := bucket.NewUserBucketClient(string(tenantID), c.bucketClient, c.cfgProvider)
+	reader := bodyReader{
+		r: r,
+	}
+	if err := bkt.Upload(ctx, dst, &reader); err != nil {
+		return errors.Wrap(err, "failed uploading block file to bucket")
+	}
+
+	level.Info(c.logger).Log("msg", "finished uploading block file to bucket", "user",
+		tenantID, "block_id", blockID, "path", pth)
+
+	return nil
+}
+
+type bodyReader struct {
+	r *http.Request
+}
+
+// ObjectSize implements thanos.ObjectSizer.
+func (r *bodyReader) ObjectSize() (int64, error) {
```

Comment:
Nit: there's no need to make this functions pointer receivers since the only thing that this struct contains is already a reference.

Reply:
Done.